File: src/registry/registry.js
import { readFileSync } from 'node:fs';
import path from 'node:path';
import ajvModule from 'ajv/dist/2020.js';
import ajvFormatsModule from 'ajv-formats';
import { sha256Text } from '../util/hash.js';

// ajv and ajv-formats are CommonJS. Node hands back the class itself, but the
// interop shape differs under type checking, so normalize once here.
const Ajv = /** @type {any} */ (ajvModule).default ?? /** @type {any} */ (ajvModule);
const addFormats =
  /** @type {any} */ (ajvFormatsModule).default ?? /** @type {any} */ (ajvFormatsModule);

/**
 * Team registry loading and validation (design section 3.1, flow step 1).
 *
 * Ownership is supplied, never inferred. Validation completes before any Elasticsearch
 * query, because a registry mistake silently changes which alerts belong to a team and
 * that error is invisible in the output.
 */

export const DEFAULT_REGISTRY_PATH = path.join('config', 'teams.json');
const SCHEMA_PATH = path.join('config', 'teams.schema.json');

/**
 * @typedef {object} PanelVariable
 * @property {string} name
 * @property {'custom'|'constant'|'interval'|'query'} type
 * @property {string} [value]
 * @property {string[]} [values]
 * @property {boolean} [multi]
 * @property {boolean} [all_selected]
 */

/**
 * @typedef {object} Panel
 * @property {string} panel_id
 * @property {'v1'|'v2'} schema
 * @property {string} sql
 * @property {PanelVariable[]} [variables]
 */

/**
 * @typedef {object} TeamEntry
 * @property {string} team_id
 * @property {string} display_name
 * @property {string[]} v1_operators
 * @property {string|null} v2_operator
 * @property {Panel[]} [panels]
 */

/**
 * @typedef {object} Registry
 * @property {string} registry_version
 * @property {string} [effective_date]
 * @property {TeamEntry[]} teams
 */

/**
 * @typedef {object} LoadedRegistry
 * @property {Registry} registry
 * @property {string} registryVersion
 * @property {string} fileSha256 SHA-256 of the complete registry document as read.
 * @property {string} filePath
 */

/** Raised for any registry problem. Callers fail the run before querying alerts. */
export class RegistryError extends Error {
  /**
   * @param {string} message
   * @param {string[]} [details]
   */
  constructor(message, details = []) {
    super(details.length ? `${message}\n  - ${details.join('\n  - ')}` : message);
    this.name = 'RegistryError';
    this.details = details;
  }
}

/**
 * Validate a parsed registry document against the checked-in JSON Schema plus the
 * cross-entry rules the schema cannot express.
 *
 * @param {unknown} doc
 * @returns {Registry}
 */
export function validateRegistry(doc) {
  const schema = JSON.parse(readFileSync(SCHEMA_PATH, 'utf8'));
  const ajv = new Ajv({ allErrors: true, strict: true });
  addFormats(ajv);
  const validate = ajv.compile(schema);

  if (!validate(doc)) {
    const details = (validate.errors || []).map(
      (e) =>
        `${e.instancePath || '/'} ${e.message}${e.params ? ` ${JSON.stringify(e.params)}` : ''}`,
    );
    throw new RegistryError('registry does not match config/teams.schema.json', details);
  }

  const registry = /** @type {Registry} */ (doc);
  /** @type {string[]} */
  const problems = [];

  /** @type {Map<string, string>} team_id -> first occurrence */
  const seenTeamIds = new Map();
  /** Exact, case-sensitive operator value -> owning team_id (design section 3.1). */
  const seenOperators = new Map();

  for (const team of registry.teams) {
    if (seenTeamIds.has(team.team_id)) {
      problems.push(`duplicate team_id "${team.team_id}"`);
    }
    seenTeamIds.set(team.team_id, team.team_id);

    // "at least one source operator must exist" — an entry with neither cannot be
    // queried at all, and silently returning zero alerts would read as a clean team.
    if (team.v1_operators.length === 0 && team.v2_operator === null) {
      problems.push(
        `team "${team.team_id}" has no source operator: v1_operators is empty and v2_operator is null`,
      );
    }

    // Uniqueness is a CROSS-team rule. One team may legitimately carry the same string
    // as both a v1 operator and its v2 operator (a team that kept its name through the
    // migration), so the same value seen twice inside one entry is not a conflict.
    const own = new Set([
      ...team.v1_operators,
      ...(team.v2_operator === null ? [] : [team.v2_operator]),
    ]);
    for (const op of own) {
      const owner = seenOperators.get(op);
      if (owner !== undefined && owner !== team.team_id) {
        problems.push(
          `operator "${op}" is claimed by both "${owner}" and "${team.team_id}" (matching is exact and case-sensitive)`,
        );
      } else {
        seenOperators.set(op, team.team_id);
      }
    }

    /** @type {Set<string>} */
    const panelIds = new Set();
    for (const panel of team.panels || []) {
      if (panelIds.has(panel.panel_id)) {
        problems.push(`team "${team.team_id}" has duplicate panel_id "${panel.panel_id}"`);
      }
      panelIds.add(panel.panel_id);

      /** @type {Set<string>} */
      const varNames = new Set();
      for (const v of panel.variables || []) {
        if (varNames.has(v.name)) {
          problems.push(
            `panel "${panel.panel_id}" of team "${team.team_id}" defines variable "${v.name}" twice`,
          );
        }
        varNames.add(v.name);
        if ((v.type === 'constant' || v.type === 'interval') && v.value === undefined) {
          problems.push(
            `panel "${panel.panel_id}" variable "${v.name}" is ${v.type} but has no "value"`,
          );
        }
        if (v.type === 'custom' && v.values === undefined) {
          problems.push(
            `panel "${panel.panel_id}" variable "${v.name}" is custom but has no "values"`,
          );
        }
      }
    }
  }

  if (problems.length) throw new RegistryError('registry validation failed', problems);
  return registry;
}

/**
 * Read, hash and validate the registry document.
 *
 * The hash covers the complete file bytes rather than the parsed object: a run must be
 * able to prove exactly which document it used, including entries for teams it did not
 * select.
 *
 * @param {string} [filePath]
 * @returns {LoadedRegistry}
 */
export function loadRegistry(filePath = DEFAULT_REGISTRY_PATH) {
  /** @type {string} */
  let raw;
  try {
    raw = readFileSync(filePath, 'utf8');
  } catch (err) {
    throw new RegistryError(
      `cannot read registry at ${filePath}: ${/** @type {any} */ (err).message}`,
    );
  }

  /** @type {unknown} */
  let parsed;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    throw new RegistryError(
      `registry at ${filePath} is not valid JSON: ${/** @type {any} */ (err).message}`,
    );
  }

  const registry = validateRegistry(parsed);
  return {
    registry,
    registryVersion: registry.registry_version,
    fileSha256: sha256Text(raw),
    filePath,
  };
}

/**
 * Select exactly one team. A run never defaults to all teams (design section 6).
 *
 * @param {LoadedRegistry} loaded
 * @param {string} teamId
 * @returns {TeamEntry}
 */
export function selectTeam(loaded, teamId) {
  const team = loaded.registry.teams.find((t) => t.team_id === teamId);
  if (!team) {
    const known = loaded.registry.teams
      .map((t) => t.team_id)
      .sort()
      .join(', ');
    throw new RegistryError(`team "${teamId}" is not in the registry. Known teams: ${known}`);
  }
  return team;
}

/**
 * Immutable snapshot of the selected entry, stored with the run so the ownership used is
 * reproducible even after the registry is edited.
 *
 * @param {TeamEntry} team
 * @returns {string}
 */
export function snapshotTeamEntry(team) {
  return JSON.stringify(team);
}

File: src/alerts_bi/registry.py
"""Team registry loading and validation (design section 3.1, flow step 1).

Ownership is supplied, never inferred. Validation completes before any Elasticsearch
query, because a registry mistake silently changes which alerts belong to a team and that
error is invisible in the output.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator

from alerts_bi.hashing import sha256_text

__all__ = [
    "DEFAULT_REGISTRY_PATH",
    "SCHEMA_PATH",
    "LoadedRegistry",
    "Panel",
    "PanelVariable",
    "RegistryError",
    "TeamEntry",
    "load_registry",
    "select_team",
    "snapshot_team_entry",
    "validate_registry",
]

DEFAULT_REGISTRY_PATH = Path("config") / "teams.json"
SCHEMA_PATH = Path("config") / "teams.schema.json"


class RegistryError(Exception):
    """Raised for any registry problem, before alerts are queried."""

    def __init__(self, message: str, details: list[str] | None = None) -> None:
        self.details = details or []
        if self.details:
            message = message + "\n  - " + "\n  - ".join(self.details)
        super().__init__(message)


@dataclass(frozen=True, slots=True)
class PanelVariable:
    name: str
    type: str
    value: str | None = None
    values: list[str] | None = None
    multi: bool = False
    all_selected: bool = False

    @staticmethod
    def from_dict(raw: dict[str, Any]) -> PanelVariable:
        return PanelVariable(
            name=raw["name"],
            type=raw["type"],
            value=raw.get("value"),
            values=list(raw["values"]) if isinstance(raw.get("values"), list) else None,
            multi=bool(raw.get("multi", False)),
            all_selected=bool(raw.get("all_selected", False)),
        )


@dataclass(frozen=True, slots=True)
class Panel:
    panel_id: str
    schema: str
    sql: str
    variables: tuple[PanelVariable, ...] = ()

    @staticmethod
    def from_dict(raw: dict[str, Any]) -> Panel:
        return Panel(
            panel_id=raw["panel_id"],
            schema=raw["schema"],
            sql=raw["sql"],
            variables=tuple(PanelVariable.from_dict(v) for v in raw.get("variables", [])),
        )


@dataclass(frozen=True, slots=True)
class TeamEntry:
    team_id: str
    display_name: str
    v1_operators: tuple[str, ...]
    v2_operator: str | None
    panels: tuple[Panel, ...] = ()
    #: The entry exactly as it appeared in the registry document, preserved so the stored
    #: snapshot is the supplied text rather than a re-serialization of our own model.
    raw: dict[str, Any] | None = None

    @staticmethod
    def from_dict(raw: dict[str, Any]) -> TeamEntry:
        return TeamEntry(
            team_id=raw["team_id"],
            display_name=raw["display_name"],
            v1_operators=tuple(raw["v1_operators"]),
            v2_operator=raw["v2_operator"],
            panels=tuple(Panel.from_dict(p) for p in raw.get("panels", [])),
            raw=raw,
        )

    def panels_for(self, schema: str) -> list[Panel]:
        """Panels that can describe rows of one schema.

        A v2 panel cannot exclude a v1 row, so suppression is always evaluated within a
        schema.
        """
        return [panel for panel in self.panels if panel.schema == schema]


@dataclass(frozen=True, slots=True)
class LoadedRegistry:
    registry: dict[str, Any]
    registry_version: str
    #: SHA-256 of the complete registry document as read, covering entries for teams this
    #: run did not select.
    file_sha256: str
    file_path: Path
    teams: tuple[TeamEntry, ...]


def validate_registry(document: Any, schema_path: Path = SCHEMA_PATH) -> tuple[TeamEntry, ...]:
    """Validate against the checked-in JSON Schema plus the cross-entry rules.

    The schema cannot express the rules that span entries, and those are precisely the
    ones that corrupt attribution: an operator claimed by two teams, or an entry with no
    source operator at all.
    """
    schema = json.loads(schema_path.read_bytes().decode("utf-8"))
    validator = Draft202012Validator(schema)
    errors = sorted(validator.iter_errors(document), key=lambda e: list(e.absolute_path))
    if errors:
        details = [
            f"{'/' + '/'.join(str(p) for p in e.absolute_path) if e.absolute_path else '/'} "
            f"{e.message}"
            for e in errors
        ]
        raise RegistryError(f"registry does not match {schema_path.as_posix()}", details)

    problems: list[str] = []
    seen_team_ids: set[str] = set()
    # Exact, case-sensitive operator value -> owning team_id.
    seen_operators: dict[str, str] = {}
    entries: list[TeamEntry] = []

    for raw_team in document["teams"]:
        team = TeamEntry.from_dict(raw_team)
        entries.append(team)

        if team.team_id in seen_team_ids:
            problems.append(f'duplicate team_id "{team.team_id}"')
        seen_team_ids.add(team.team_id)

        # "At least one source operator must exist": an entry with neither cannot be
        # queried at all, and silently returning zero alerts would read as a clean team.
        if not team.v1_operators and team.v2_operator is None:
            problems.append(
                f'team "{team.team_id}" has no source operator: '
                "v1_operators is empty and v2_operator is null"
            )

        # Uniqueness is a CROSS-team rule. One team may legitimately carry the same string
        # as both a v1 operator and its v2 operator - a team that kept its name through
        # the migration - so the same value twice inside one entry is not a conflict.
        own = set(team.v1_operators)
        if team.v2_operator is not None:
            own.add(team.v2_operator)
        for operator in sorted(own):
            owner = seen_operators.get(operator)
            if owner is not None and owner != team.team_id:
                problems.append(
                    f'operator "{operator}" is claimed by both "{owner}" and '
                    f'"{team.team_id}" (matching is exact and case-sensitive)'
                )
            else:
                seen_operators[operator] = team.team_id

        panel_ids: set[str] = set()
        for panel in team.panels:
            if panel.panel_id in panel_ids:
                problems.append(f'team "{team.team_id}" has duplicate panel_id "{panel.panel_id}"')
            panel_ids.add(panel.panel_id)

            variable_names: set[str] = set()
            for variable in panel.variables:
                if variable.name in variable_names:
                    problems.append(
                        f'panel "{panel.panel_id}" of team "{team.team_id}" defines '
                        f'variable "{variable.name}" twice'
                    )
                variable_names.add(variable.name)
                if variable.type in ("constant", "interval") and variable.value is None:
                    problems.append(
                        f'panel "{panel.panel_id}" variable "{variable.name}" is '
                        f'{variable.type} but has no "value"'
                    )
                if variable.type == "custom" and variable.values is None:
                    problems.append(
                        f'panel "{panel.panel_id}" variable "{variable.name}" is custom '
                        'but has no "values"'
                    )

    if problems:
        raise RegistryError("registry validation failed", problems)
    return tuple(entries)


def load_registry(
    file_path: Path | str = DEFAULT_REGISTRY_PATH,
    schema_path: Path = SCHEMA_PATH,
) -> LoadedRegistry:
    """Read, hash and validate the registry document.

    The hash covers the complete file BYTES rather than the parsed object, so a run can
    prove exactly which document it used - including the entries for teams it did not
    select. Bytes are decoded explicitly rather than read through universal-newline
    translation, which would silently change the hash on a CRLF checkout.
    """
    path = Path(file_path)
    try:
        raw_bytes = path.read_bytes()
    except OSError as exc:
        raise RegistryError(f"cannot read registry at {path.as_posix()}: {exc}") from exc

    raw_text = raw_bytes.decode("utf-8")
    try:
        document = json.loads(raw_text)
    except json.JSONDecodeError as exc:
        raise RegistryError(f"registry at {path.as_posix()} is not valid JSON: {exc}") from exc

    teams = validate_registry(document, schema_path)
    return LoadedRegistry(
        registry=document,
        registry_version=document["registry_version"],
        file_sha256=sha256_text(raw_text),
        file_path=path,
        teams=teams,
    )


def select_team(loaded: LoadedRegistry, team_id: str) -> TeamEntry:
    """Select exactly one team. A run never defaults to all teams (design section 6)."""
    for team in loaded.teams:
        if team.team_id == team_id:
            return team
    known = ", ".join(sorted(t.team_id for t in loaded.teams))
    raise RegistryError(f'team "{team_id}" is not in the registry. Known teams: {known}')


def snapshot_team_entry(team: TeamEntry) -> str:
    """Immutable snapshot of the selected entry, stored with the run.

    Serialized compactly and without ASCII escaping so the stored text is byte-identical
    to what the superseded JavaScript implementation stored, which keeps run records
    comparable across the port.
    """
    source = team.raw if team.raw is not None else _entry_to_dict(team)
    return json.dumps(source, separators=(",", ":"), ensure_ascii=False)


def _entry_to_dict(team: TeamEntry) -> dict[str, Any]:
    entry: dict[str, Any] = {
        "team_id": team.team_id,
        "display_name": team.display_name,
        "v1_operators": list(team.v1_operators),
        "v2_operator": team.v2_operator,
    }
    if team.panels:
        entry["panels"] = [
            {"panel_id": p.panel_id, "schema": p.schema, "sql": p.sql} for p in team.panels
        ]
    return entry
